components/fonts/shapers/harfrust.rs

```diff
@@ -0,0 +1,159 @@
+                *y_pos -= y_advance;
+            }
+
+            Some(Point2D::new(x_offset, *y_pos - y_offset))
+        };
+
+        ShapedGlyphEntry {
+            codepoint: glyph_info_i.glyph_id,
+            advance: x_advance,
+            offset,
+        }
+    }
+}
+
+pub struct Shaper {
+    font: *const Font,
+    /// The raw byte data of the font
+    font_data: FontData,
+    /// The index of a font in it's collection (.ttc)
+    /// If the font file is not a collection then this is 0
+    font_index: u32,
+    // points-per-em. Used for scaling HarfRust's output
+    ppem: f64,
+}
+
+// `Font` and `FontData` are both threadsafe, so we can make the data structures here as thread-safe as well.
+#[allow(unsafe_code)]
+unsafe impl Sync for Shaper {}
+#[allow(unsafe_code)]
+unsafe impl Send for Shaper {}
+
+impl Shaper {
+    pub(crate) fn new(font: &Font) -> Self {
+        let font_data = font.data().clone();
+        let font_index = font.identifier().index();
+        // Set points-per-em. if zero, performs no hinting in that direction
+        let ppem = font.descriptor.pt_size.to_f64_px();
+        Self {
+            font: font as *const Font,
+            font_data,
+            font_index,
+            ppem
+        }
+    }
+}
+
+impl THarfShaper for Shaper {
+    type ShapedGlyphData = ShapedGlyphData;
+
+    fn shape_text(&self, text: &str, options: &crate::ShapingOptions) -> ShapedGlyphData {
+        let mut buffer = UnicodeBuffer::new();
+
+        // Set direction
+        buffer.set_direction(if options.flags.contains(ShapingFlags::RTL_FLAG) {
+            harfrust::Direction::RightToLeft
+        } else {
+            harfrust::Direction::LeftToRight
+        });
+
+        // Set script
+        let script =
+            Script::from_iso15924_tag(Tag::from_u32(unicode_to_hb_script(options.script))).unwrap();
+        buffer.set_script(script);
+
+        // Push text
+        buffer.push_str(text);
+
+        // Features
+        let mut features = SmallVec::<[Feature; 2]>::new();
+        if options
+            .flags
+            .contains(ShapingFlags::IGNORE_LIGATURES_SHAPING_FLAG)
+        {
+            features.push(Feature::new(Tag::new(b"liga"), 0, ..));
+        }
+        if options
+            .flags
+            .contains(ShapingFlags::DISABLE_KERNING_SHAPING_FLAG)
+        {
+            features.push(Feature::new(Tag::new(b"kern"), 0, ..));
+        }
+
+        let hr_font =
+            HarfRustFontRef::from_index(self.font_data.as_ref(), self.font_index).unwrap();
+        let shaper_data = ShaperData::new(&hr_font);
+
+        // TODO: handle font variations
+        // let variations: Vec<Variation> = Vec::new();
+        // let instance_data = ShaperInstance::from_variations(&hr_font, &variations);
+
+        let shaper = shaper_data
+            .shaper(&hr_font)
+            // Set the instance
+            // .instance(Some(&instance_data))
+            .point_size(Some(self.ppem as f32))
```

Comment:
This is only for applying the tracking table and doesn’t actually enable scaling. I’d recommend not setting this value for now and I’ll adjust the API to make this more clear.

Reply:
That makes sense, but do we not want the tracking table to be applied?

Reply:
We do and the default tracking will be applied unless the (pseudo) `trak` feature is disabled. This is for selecting the _point_ (not pixel) size used to look up the proper adjustment in the table. The equivalent HarfBuzz function is `hb_font_set_ptem` which I don't see called in the HB shaper code here.

If you'd like to set this, feel free to read the protracted discussion on _which_ value to submit here: https://github.com/harfbuzz/harfbuzz/pull/1484 :)

Reply:
Hmmm.... yeah, maybe I'll leave this one for now!